### PR TITLE
Fix for file descriptor leak.

### PR DESCRIPTION
**Pull request by jpack**

### CODE REVIEW

File: src/main/java/org/devnull/jedi/configs/JediConfig.java [hunk 1]
```diff
@@ -72,6 +72,11 @@
 	public String unix_socket_path = null;
 
 	/**
+	 * Timeout value for the unix socket. Without this sockets get stuck in the CLOSE_WAIT status and file descriptors leak.
```
Comment (dhawth) [inline src/main/java/org/devnull/jedi/configs/JediConfig.java]: When adding javadoc for time-related parameters, always include the unit of measure.  This should explicitly say "timeout in milliseconds".

File: src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java [hunk 2]
```diff
@@ -125,7 +126,14 @@ public void run()
 					log.debug("waiting to read request from socket");
 				}
 
-				requestLine = reader.readLine();
+				try {
+					requestLine = reader.readLine();
+				} catch (IOException e){
```
Comment (dhawth) [inline src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java:130]: This should be SocketTimeoutException

File: src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java [hunk 3]
```diff
@@ -125,7 +126,14 @@ public void run()
...
+					//expected, PowerDNS doesn't tell us when it's done using a connection, we need a timeout to clean up correctly.
+					log.debug("socket timed out, closing");
+					break;
```
Comment (dhawth) [inline src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java:135]: Any reason we shouldn't just continue?

File: src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java [hunk 4]
```diff
@@ -125,7 +126,14 @@ public void run()
 					log.debug("waiting to read request from socket");
 				}
 
-				requestLine = reader.readLine();
+				try {
+					requestLine = reader.readLine();
+				} catch (IOException e){
+					//expected, PowerDNS doesn't tell us when it's done using a connection, we need a timeout to clean up correctly.
+					log.debug("socket timed out, closing");
```
Comment (dhawth) [inline src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java]: We're not closing the socket here, but we can log the timeout.  We should also emit a statsd counter for it.

File: src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java [hunk 5]
```diff
@@ -74,6 +74,7 @@ public PowerDNSConnectionHandler(Socket client,
 		this.apiPool = apiPool;
 		this.cache = cache;
 		restClient = new RestClient(config);
+		socket.setSoTimeout(config.unix_socket_timeout);
```
Comment (dhawth) [inline src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java]: You're configuring every socket with the _unix_ socket timeout here.  Shouldn't that be done in the UnixSocketServer class in Jedi instead?  The PowerDNSConnectionHandler handles **all** sockets.

File: src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java (hunk 3 above)
Comment (jpack) [inline src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java:135]: Yeah, we'd just get stuck reading from the socket again. The crux of the problem seems to be that the socket expects more data even after PowerDNS is done sending requests and the TCP connection is closed.

File: src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java (hunk 5 above)
Comment (jpack) [inline src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java]: Yeah, good catch.

File: src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java (hunk 2 above)
Comment (jpack) [inline src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java:130]: Actually the exception that bubbles up here is an IOException from the unix socket library.

**Pull request merged**